L1Trigger/L1THGCal/src/backend_emulator/HGCalCluster_SA.cc

```diff
@@ -1,7 +1,245 @@
 #include "L1Trigger/L1THGCal/interface/backend_emulator/HGCalCluster_SA.h"
+#include <math.h>
+#include <algorithm>
+
 
 using namespace l1thgcfirmware;
 
+double HGCalCluster::Sigma_Energy(unsigned int N_TC_W, unsigned long int Sum_W2, unsigned int Sum_W) {
+  unsigned long int N = N_TC_W*Sum_W2 - pow(Sum_W,2);
+  unsigned long int D = pow(N_TC_W,2);
+  return  (double)sqrt(N/D);
+}
+
+double HGCalCluster::Mean_coordinate(unsigned int Sum_Wc, unsigned int Sum_W) {
+  return (double)Sum_Wc/Sum_W;
+}
+
+double HGCalCluster::Sigma_Coordinate(unsigned int Sum_W, unsigned long int Sum_Wc2, unsigned int Sum_Wc) {
+  unsigned long int N = Sum_W*Sum_Wc2 - pow(Sum_Wc,2);
+  unsigned long int D = pow(Sum_W,2);
+  return (double)sqrt(N/D);
+}
+
+double HGCalCluster::Energy_ratio(unsigned int E_N, unsigned int E_D) {
+  if ( E_D == 0 ) { 
+    return 0; 
+  } else {
+    return (double)E_N/E_D;
+  }
+}
+
+std::vector<int> HGCalCluster::ShowerLengthProperties(unsigned long int layerBits) {
+  int counter = 0;
+  int firstLayer = 0;
+  bool firstLayerFound = false;
+  int lastLayer = 0;
+ 
+
+  std::vector<int> layerBits_array;
+  for (int idx = 0; idx<36; idx++) {
+    if ( (layerBits&(1L<<(35-idx)) ) >= 1L ) {
+      if ( !firstLayerFound ) {
+        firstLayer = idx+1;
+        firstLayerFound=true;
+      }
+      lastLayer = idx+1;
+      counter += 1;
+    } else {
+      layerBits_array.push_back(counter);
+      counter = 0;
+    }
+  }
+  int showerLen = lastLayer - firstLayer + 1;
+  int coreShowerLen = 36;
+  if ( layerBits_array.size()>0 ) {
+    coreShowerLen = *std::max_element(layerBits_array.begin(), layerBits_array.end()); 
+  }
+
+  std::vector<int> output = {firstLayer,lastLayer, showerLen, coreShowerLen};
+
+  return output;
+}
+
+
+unsigned long int HGCalCluster::Sigma_E_Fraction()  { 
+  double Sigma_E_temp = Sigma_Energy( this->n_tc_w(), this->w2(), this->w() );
+  double intpart;
+  return modf(Sigma_E_temp,&intpart)*pow(2,1);
+}
+
+
+unsigned long int HGCalCluster::Sigma_E_Quotient()  { 
+  double Sigma_E_temp = Sigma_Energy( this->n_tc_w(), this->w2(), this->w() );
+  double intpart;
+  double frac =  modf(Sigma_E_temp,&intpart);
+  return intpart;
+}
```

Comment:
Could this be combined into one function with `Sigma_E_Fraction` that returns a `std::pair`?  Similar for the other Fraction/Quotient functions.  Or would this deviate from the python emulator/firmware?

Reply:
done in 342389f